src/mqtt_client.rs

```diff
@@ -437,6 +450,12 @@ impl<'buf, TcpStack: TcpClientStack, Clock: embedded_time::Clock>
             return Ok(());
         }
 
+        if let Some(max) = &self.sm.context().max_qos {
+            if self.downgrade_qos && publish.qos > *max {
```

Comment:
I think we'll also need to apply this to the `Will`.

Reply:
The will is set before the maximum QoS is known (it's sent as part of the CONNECT payload and we don't learn the max QoS until the CONNACK). I assume the broker will handle capping its QoS